File: Sorts.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
public class Sorts {
    private static Tone tone;
    private static int comparisons;
    /**
     * Initializes Sort with a @param tone instance to play tones while sorting.    
     */
    public Sorts(Tone t) {
        tone = t;
        comparisons = 0;
    }

    public void bubbleSort(double[] nums) {
        comparisons = 0;
        for (int i = 0; i < nums.length; i++) {
            for (int j = 0; j < nums.length - i - 1; j++) {
                double cur = nums[j];
                double next = nums[j + 1];
                double mag = Math.abs(cur - next);
                //mag = (cur + next) / 2;
                comparisons++;
                tone.generateSound(mag);
                if (cur > next) {
                    nums[j] = next;
                    nums[j + 1] = cur;
                }
            }
        }
        System.out.println("Comparisons: " + comparisons);
    }

    public void insertionSort(double[] nums) {
        comparisons = 0;
        double min = nums[0];
        int index = 0;
        for (int i = 1; i < nums.length; i++) {
            comparisons++;
            double mag = Math.abs(nums[i] - min);
            tone.generateSound(mag);
            if (nums[i] < min) {
                min = nums[i];
                index = i;
            }
        }
        double zero = nums[0];
        nums[0] = min;
        nums[index] = zero;
        for (int i = 2; i < nums.length; i++) {
            int j = i;
            
            while (j > 0 && nums[j-1] > nums[j]) {
                comparisons++;
                double next = nums[j - 1];
                double mag = Math.abs(nums[j] - next);
                tone.generateSound(mag);
                nums[j - 1] = nums[j];
                nums[j] = next;
                j--;
            }
        }

        System.out.println("Comparisons: " + comparisons);
    }

    public void mergeSort(double[] nums) {
        comparisons = 0;
        ms(nums, 0, nums.length - 1);
        System.out.println("Comparisons: " + comparisons);
    }

    private static void merge(double[] nums, int low, int mid, int high) {
        int left_size = mid - low + 1;
        int right_size = high - mid;
        double[] left = new double[left_size];
        double[] right = new double[right_size];
        for (int i = 0; i < left_size; i++) {
            left[i] = nums[low + i];
        }
        for (int i = 0; i < right_size; i++) {
            right[i] = nums[mid + 1 + i];
        }
        int l = 0, r = 0;
        int k = low;
        while (l < left_size && r < right_size) {
            double mag = Math.abs(left[l] - right[r]);
            tone.generateSound(mag);
            if (left[l] <= right[r]) {
                nums[k] = left[l];
                l++;
            } else {
                nums[k] = right[r];
                r++;
            }
            k++;
            comparisons++;
        }

        while (l < left_size) {
            nums[k] = left[l];
            l++;
            k++;
        }

        while(r < right_size) {
            nums[k] = right[r];
            r++;
            k++;
        }
    }

    private static void ms(double[] nums, int low, int high) {
        if (low < high) {
            int mid = (low + high - 1) / 2;
            ms(nums, low, mid);
            ms(nums, mid + 1, high);

            merge(nums, low, mid, high);
        }
    }

    public void quickSort(double[] nums) {
        comparisons = 0;
        qs(nums, 0, nums.length - 1);
        System.out.println("Comparisons: " + comparisons);
    }

    private static void qs(double[] nums, int left, int right) {
        if (left >= right) {
            return;
        }
        int pivot = partition(nums, left, right);
        qs(nums, left, pivot - 1);
        qs(nums, pivot + 1, right);
    }

    private static int partition(double[] nums, int left, int right) {
        double pivot = nums[right];
        int i = left - 1;
        for (int j = left; j < right; j++) {
            comparisons++;
            double mag = Math.abs(nums[j] - pivot);
            tone.generateSound(mag);
            if (nums[j] < pivot) {
                i++;
                double temp = nums[j];
                nums[j] = nums[i];
                nums[i] = temp;
            }
        }
        double temp = nums[i + 1];
        nums[i + 1] = nums[right];
        nums[right] = temp;
        return i + 1; 
    }

    public void LSDRadixSort(double[] nums) {
        // LSD radix
        double m = getMax(nums);
        int D = (int) (Math.log(m + 1) / Math.log(4));
        double[] temp = new double[nums.length];
        int d = 0;
        while (d < D) {
            double[] count = new double[10];
            for (int i = 0; i < count.length; i++) {
                count[i] = 0;
            }
            for (int i = 0; i < nums.length; i++) {
                count[(int)(nums[i] / (Math.pow(10, d))) % 10]++;
            }
            // Give count[i] the correct starting index based on counts found in last loop
            for (int i = 1; i < 10; i++) {
                count[i] += count[i - 1];
            }
            for (int i = nums.length - 1; i >= 0; i--) {
              temp[(int) count[(int)(nums[i] / (Math.pow(10, d))) % 10] - 1] = nums[i];
              count[(int)(nums[i] / (Math.pow(10, d))) % 10]--;
            }
            for (int i = 0; i < nums.length; i++) {
                tone.generateSound(temp[i]);
                nums[i] = temp[i];
            }
            d++;
        }

    }

    private static double getMax(double[] nums) {
        double m = nums[0];
        for (int i = 1; i < nums.length; i++) {
            if (nums[i] > m) {
                m = nums[i];
            }
        }
        return m;
    }
    
    public void MSDRadixSort(double[] nums) {
        double m = getMax(nums);
        int D = (int) (Math.log10(m) + 1);
        System.out.println(D);
        msdr(nums, 0, nums.length - 1, D);
    }

    
    private static void msdr(double[] nums, int start, int end, int D) {
        //for (int i = 0; i < nums.length; i++) {
            //System.out.print(nums[i] + " ");
        //}
        //System.out.println("\n" + start + " " + end);
        if (start + 1 >= end || D < 1) return;
        int[] count = new int[11];
        Arrays.fill(count, 0);
        for (int i = 0; i < nums.length; i++) {
            int digit = ((int) (nums[i] / Math.pow(10, D - 1))) % 10;
            count[digit + 1]++;
        }
        for (int i = 0; i < 10; i++) {
            count[i + 1] += count[i]; // starting index for each 
        }
        double[] temp = new double[nums.length];
        for (int i = start; i <= end; i++) {
            int digit = (int) (nums[i] / Math.pow(10, D - 1) % 10);
            temp[count[digit]++] = nums[i];
        }
        for (int i = 0; i < nums.length; i++) {
            nums[i] = temp[i];
        }
        for (int i = 0; i < 10; i++) {
            msdr(nums, start + count[i], start + count[i + 1], D - 1);
        }

        for (int i = 0; i < nums.length; i++) {
            System.out.print(nums[i] + " ");
        }
        System.out.println("\n");
    }
}

File: sorts.js


let comparisons = 0;
/*
function bubbleSort(nums) {
    comparisons = 0;
    for (var i = 0; i < nums.length; i++) {
        for (var j = 0; j < nums.length - i - 1; j++) {
            cur = nums[j];
            next = nums[j + 1];
            mag = Math.abs(cur - next);
            comparisons++;
            //tone.generateSound(mag);
            if (cur > next) {
                nums[j] = next;
                nums[j + 1] = cur;
            }
        }
    }
    console.log(nums);
    console.log(comparisons);
    document.getElementById("stuff").innerHTML = nums;
}*/
function bubbleSort(nums) {
    document.getElementById("header").innerHTML = "Bubble Sort Selected";
    hideButtons("bubble");
    comparisons = 0;
    for (var i = 0; i < nums.length; i++) {
        for (var j = 0; j < nums.length - i - 1; j++) {
            cur = nums[j];
            next = nums[j + 1];
            //mag = Math.abs(cur - next);
            comparisons++;
            //tone.generateSound(mag);
            if (cur > next) {
                let temp = colorArr[j + 1];
                colorArr[j + 1] = colorArr[j];
                colorArr[j] = temp;
                nums[j] = next;
                nums[j + 1] = cur;
                //clear();
                //sleep(.1);
                redraw();
                return;
            }
        }
    }
    // if we make it here we are done sorting
    sorted = 1;
    console.log(nums);
    console.log(comparisons);
    document.getElementById("stuff").innerHTML = nums;
}


function insertionSort(nums) {
    document.getElementById("header").innerHTML = "Insertion Sort Selected";
    hideButtons("insertion");
    comparisons = 0;
    let min = nums[0];
    let index = 0;
    for (var i = 1; i < nums.length; i++) {
        comparisons++;
        mag = Math.abs(nums[i] - min);
        if (nums[i] < min) {
            min = nums[i];
            index = i;
        }
    }
    let zero = nums[0];
    nums[0] = min;
    nums[index] = zero;
    for (var i = 2; i < nums.length; i++) {
        let j = i;
        while (j > 0 && nums[j - 1] > nums[j]) {
            comparisons++;
            let next = nums[j - 1];
            let mag = Math.abs(nums[j] - next);
            nums[j - 1] = nums[j];
            nums[j] = next;
            let cNext = colorArr[j - 1];
            colorArr[j - 1] = colorArr[j];
            colorArr[j] = cNext;
            redraw();
            j--;
            return;
        }
    }

    console.log(nums);
    console.log(comparisons);
    document.getElementById("stuff").innerHTML = nums;
}


// may need to do iterative mergesort to work for my vis?
function mergeSort(nums) {
    document.getElementById("header").innerHTML = "Merge Sort Selected";
    hideButtons("merge");
    comparisons = 0;
    nums = ms(nums);
    console.log(comparisons);
    console.log(nums);
    document.getElementById("stuff").innerHTML = nums;
}

function ms(nums) {
    if (nums.length < 2) {
        return nums;
    }
    let mid = Math.floor(nums.length / 2);
    let first = ms(nums.slice(0, mid));
    let second = ms(nums.slice(mid));

    return merge(first, second);
}

function merge(first, second) {
    let arr = [];
    while (first.length && second.length) {
        comparisons++;
        if (first[0] < second[0]) {
            arr.push(first.shift());
            //first.shift();
        } else {
            arr.push(second.shift());
            //second.shift();
        }
    }
    return [...arr, ...first, ...second];
}

// may need to do iterative quicksort to work for my vis?
function quickSort(nums) {
    document.getElementById("header").innerHTML = "Quick Sort Selected";
    hideButtons("quick");
    comparisons = 0;
    qs(nums, 0, nums.length - 1);
    console.log(comparisons);
    console.log(nums);
    document.getElementById("stuff").innerHTML = nums;
}

function qs(nums, left, right) {
    if (left >= right) {
        return;
    }
    const pivot = partition(nums, left, right);
    qs(nums, left, pivot - 1);
    qs(nums, pivot + 1, right);
}

function partition(nums, left, right) {
    const pivot = nums[right];
    let i = left - 1;
    for (var j = left; j < right; j++) {
        comparisons++;
        let mag = Math.abs(nums[j] - pivot);
        //tone.generateSound(mag);
        if (nums[j] < pivot) {
            i++;
            let temp = nums[j];
            nums[j] = nums[i];
            nums[i] = temp;
            let cTemp = colorArr[j];
            colorArr[j] = colorArr[i];
            colorArr[i] = cTemp;
            //redraw();
            //return;
        }
    }
    let temp = nums[i + 1];
    nums[i + 1] = nums[right];
    nums[right] = temp;
    let cTemp = colorArr[i + 1];
    colorArr[i + 1] = colorArr[right];
    colorArr[right] = cTemp;
    //redraw();
    return i + 1; 
}

let d = 0;
function LSDRadixSort(nums) {
    document.getElementById("header").innerHTML = "LSD Radix Sort Selected";
    hideButtons("radix");
    const m = getMax(nums);
    const D = (int) (Math.log(m + 1) / Math.log(4));
    let temp = [];
    let tempColor = [];
    //let d = 0;
    while (d < D) {
        let count = [];
        for (var i = 0; i < 10; i++) {
            count.push(0);
            tempColor.push([0, 0, 0]);
        }
        for (var i = 0; i < nums.length; i++) {
            count[(int)(nums[i] / (Math.pow(10, d))) % 10]++;
        }
        // Give count[i] the correct starting index based on counts found in last loop
        for (var i = 1; i < 10; i++) {
            count[i] += count[i - 1];
        }
        for (var i = nums.length - 1; i >= 0; i--) {
            temp[count[(int)(nums[i] / (Math.pow(10, d))) % 10] - 1] = nums[i];
            tempColor[count[(int)(nums[i] / (Math.pow(10, d))) % 10] - 1] = colorArr[i];
            count[(int)(nums[i] / (Math.pow(10, d))) % 10]--;
        }
        for (var i = 0; i < nums.length; i++) {
            //tone.generateSound(temp[i]);
            nums[i] = temp[i];
            colorArr[i] = tempColor[i];
        }
        d++;
        redraw();
        //console.log(d);
        return;
    }
    d = 0;
    console.log(nums);
    document.getElementById("stuff").innerHTML = nums;
}

function getMax(nums) {
    let m = nums[0];
    for (var i = 1; i < nums.length; i++) {
        if (nums[i] > m) {
            m = nums[i];
        }
    }
    return m;
}